lib/Parse/ParseType.cpp

```diff
@@ -1612,71 +1605,65 @@ bool Parser::canParseSimpleTypeIdentifier() {
   consumeToken();
 
   // Parse an optional generic argument list.
-  if (startsWithLess(Tok))
-    if (!canParseGenericArguments())
-      return false;
+  if (startsWithLess(Tok) && !canParseGenericArguments())
+    return false;
 
   return true;
 }
 
 bool Parser::canParseTypeIdentifier() {
-  while (true) {
+  for (;;) {
     if (!canParseSimpleTypeIdentifier())
       return false;
 
     // Treat 'Foo.<anything>' as an attempt to write a dotted type
     // unless <anything> is 'Type' or 'Protocol'.
-    if ((Tok.is(tok::period) || Tok.is(tok::period_prefix)) &&
-        !peekToken().isContextualKeyword("Type") &&
-        !peekToken().isContextualKeyword("Protocol")) {
-      consumeToken();
-    } else {
+    if (!(Tok.is(tok::period) || Tok.is(tok::period_prefix)) ||
```

Comment:
Two reasons to restore the original version here: (1) this version no longer matches the logic that's described in the comment above it, but rather inverts it; (2) the same expression found in the original version is used elsewhere, so it is easy to see at a glance that it's testing for the same conditions as in those other places, whereas it is much more difficult for a reader to appreciate that the inverted logic here has the same intended effect. In fact, I cannot at a glance be sure that the logic has been inverted correctly.

For both these reasons, I'd strongly suggest restoring the original version.